core/src/preferences/PreferencesBackend.cpp

```diff
@@ -183,87 +183,123 @@ namespace ledger {
 
         std::string PreferencesBackend::createNewSalt(const std::shared_ptr<api::RandomNumberGenerator>& rng) {
             auto bytes = rng->getRandomBytes(128);
-            return std::string(std::begin(bytes), std::end(bytes));
+            return std::string(bytes.begin(), bytes.end());
         }
 
         void PreferencesBackend::setEncryption(
             const std::shared_ptr<api::RandomNumberGenerator>& rng,
             const std::string& password
         ) {
-            // disable encryption to check whether we have a salt already persisted
-            unsetEncryption();
-
-            auto emptySalt = std::string("");
-            auto pref = getPreferences("__core");
-            auto salt = pref->getString(ENCRYPTION_SALT_KEY, emptySalt);
-
-            if (salt == emptySalt) {
-                // we don’t have a proper salt; create one and persist it for future use
-                salt = createNewSalt(rng);
-                pref->editor()->putString(ENCRYPTION_SALT_KEY, salt)->commit();
-            }
-
-            // create the AES cipher
-            _cipher = AESCipher(rng, password, salt, PBKDF2_ITERS);
+            // setting encryption is akin to resetting with an old password that is empty
+            resetEncryption(rng, "", password);
         }
 
         void PreferencesBackend::unsetEncryption() {
             _cipher = Option<AESCipher>::NONE;
         }
 
-        void PreferencesBackend::resetEncryption(
+        bool PreferencesBackend::resetEncryption(
             const std::shared_ptr<api::RandomNumberGenerator>& rng,
             const std::string& oldPassword,
             const std::string& newPassword
         ) {
-            // turn on encryption with the old password first
-            setEncryption(rng, oldPassword);
+            Option<AESCipher> noCipher;
+            auto newCipher = noCipher;
+            auto salt = getEncryptionSalt();
+
+            if (oldPassword.empty()) {
```

Comment:
This nasting make me crazy :)

Reply:
To be more specific, I think finding cipher for new and old password are independent from each other.

Reply:
In the first place, I had independent blocks for old / new but I switched to this nesting for the reason that it’s less code and encapsulates the logic more tightly.